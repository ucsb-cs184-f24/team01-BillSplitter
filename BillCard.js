import React from 'react';
import { View, Text, TouchableOpacity, StyleSheet } from 'react-native';
import { Feather } from '@expo/vector-icons';

const BillCard = ({ item, onPress }) => {
  return (
    <TouchableOpacity onPress={onPress} activeOpacity={0.7}>
      <View style={[
        styles.billCard,
        { backgroundColor: `${item.category.color}15` },
        { borderColor: item.category.color }
      ]}>
        <View style={[
          styles.categoryIcon,
          { backgroundColor: `${item.category.color}30` }
        ]}>
          <Feather
            name={item.category.icon}
            size={24}
            color={item.category.color}
          />
        </View>
        
        <View style={styles.billContent}>
          <View style={styles.billHeader}>
            <Text style={styles.billTitle}>{item.title}</Text>
            <Text style={styles.amount}>
              ${item.amount.toFixed(2)}
            </Text>
          </View>
          
          <Text style={styles.date}>
            {new Date(item.createdAt).toLocaleDateString()}
          </Text>
        </View>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  billCard: {
    padding: 15,
    borderRadius: 12,
    marginBottom: 15,
    borderWidth: 1,
    flexDirection: 'row',
  },
  categoryIcon: {
    width: 40,
    height: 40,
    borderRadius: 20,
    justifyContent: 'center',
    alignItems: 'center',
    marginRight: 12,
  },
  billContent: {
    flex: 1,
  },
  billHeader: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginBottom: 8,
  },
  billTitle: {
    fontSize: 18,
    fontWeight: 'bold',
    flex: 1,
    marginRight: 8,
  },
  amount: {
    fontSize: 18,
    fontWeight: 'bold',
  },
  date: {
    fontSize: 12,
    color: '#999',
  },
});

export default BillCard;