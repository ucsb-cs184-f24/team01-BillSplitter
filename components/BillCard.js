import React from 'react';
import { View, Text, TouchableOpacity, StyleSheet } from 'react-native';
import { Feather } from '@expo/vector-icons';
import { categories } from './CategorySelector';
import VenmoLinker from './VenmoLinker';

const BillCard = ({ 
  item, 
  currentUser, 
  onPress, 
  onPaymentPress,
  creatorId 
}) => {
  const getCategoryColor = (categoryId) => {
    const category = categories.find(cat => cat.id === categoryId);
    return category ? category.color : '#6B7280';
  };

  const getCategoryIcon = (category) => {
    const categoryItem = categories.find(cat => cat.id === category);
    return categoryItem ? categoryItem.icon : 'grid';
  };

  const renderStatus = () => {
    if (item.isCreator) {
      if (item.totalPending > 0) {
        if (item.totalPending === 1) {
          return `$${item.pendingAmounts[0].toFixed(2)} pending from ${item.pendingUsers[0]}`;
        }
        return `${item.totalPending} pending`;
      }
      return 'Received';
    }
    if (item.status === 'paid') {
      const date = item.paidAt?.toDate();
      return `Paid${date ? ` on ${date.toLocaleDateString()}` : ''}`;
    }
    return 'You owe';
  };

  const getStatusColor = () => {
    if (item.isCreator) {
      return item.totalPending > 0 ? '#FF9500' : '#34C759';
    }
    return item.status === 'paid' ? '#34C759' : '#FF3B30';  
  };

  const getPendingText = (totalPending, pendingUsers) => {
    if (totalPending === 0) return '';
    if (totalPending === 1) return `Pending from ${pendingUsers[0]}`;
    return `Pending from ${totalPending} people`;
  };

  return (
    <TouchableOpacity 
      onPress={onPress}
      activeOpacity={0.7}
    >
      <View style={[
        styles.paymentCard,
        { backgroundColor: `${getCategoryColor(item.category)}15` },
        { borderColor: getCategoryColor(item.category) }
      ]}>
        <View style={[
          styles.categoryIcon,
          { backgroundColor: `${getCategoryColor(item.category)}30` }
        ]}>
          <Feather
            name={getCategoryIcon(item.category)}
            size={24}
            color={getCategoryColor(item.category)}
          />
        </View>
        
        <View style={styles.paymentContent}>
          <View style={styles.paymentHeader}>
            <Text style={styles.billTitle}>{item.billTitle}</Text>
            <Text style={[
              styles.amount,
              !item.isCreator && item.status === 'pending' && styles.amountOwed
            ]}>
              ${item.amount.toFixed(2)}
            </Text>
          </View>
          
          {item.billDescription ? (
            <Text style={styles.description} numberOfLines={2}>
              {item.billDescription}
            </Text>
          ) : null}
          
          <View style={styles.paymentFooter}>
            <Text style={styles.date}>
              Creation Date: {item.createdAt.toLocaleDateString()}
            </Text>
            <Text style={[styles.status, { color: getStatusColor() }]} numberOfLines={2}>
              {renderStatus()}
            </Text>
          </View>
          
          <Text style={styles.createdBy}>
            Creator: {item.creatorName}
          </Text>
          
          {!item.isCreator && item.status === 'pending' && (
            <View style={styles.buttonContainer}>
              {item.creatorVenmoUsername && (
                <View style={styles.venmoWrapper}>
                  <VenmoLinker
                    recipientId={item.creatorVenmoUsername}
                    buttonText="Pay with Venmo"
                    amount={!item.isCreator && item.status === 'pending' ? item.amount : null}
                  />
                </View>
              )}
              <TouchableOpacity
                style={[styles.payButton, { backgroundColor: getCategoryColor(item.category) }]}
                onPress={(e) => {
                  e.stopPropagation();
                  onPaymentPress(item.id);
                }}
              >
                <Text style={styles.payButtonText}>Mark as Paid</Text>
              </TouchableOpacity>
            </View>
          )}
        </View>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
    paymentCard: {
      padding: 15,
      borderRadius: 12,
      marginBottom: 15,
      borderWidth: 1,
      flexDirection: 'row',
    },
    categoryIcon: {
      width: 40,
      height: 40,
      borderRadius: 20,
      justifyContent: 'center',
      alignItems: 'center',
      marginRight: 12,
    },
    paymentContent: {
      flex: 1,
    },
    paymentHeader: {
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center',
      marginBottom: 8,
    },
    billTitle: {
      fontSize: 18,
      fontWeight: 'bold',
      flex: 1,
      marginRight: 8,
    },
    amount: {
      fontSize: 18,
      fontWeight: 'bold',
    },
    amountOwed: {
      color: '#FF3B30',  
    },
    description: {
      fontSize: 14,
      color: '#666',
      marginBottom: 8,
    },
    paymentFooter: {
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center',
      marginBottom: 4,
    },
    createdBy: {
      fontSize: 14,
      color: '#666',
    },
    status: {
      maxWidth: 125,  
      flexWrap: 'wrap',
      fontWeight: '600',
    },
    date: {
      fontSize: 12,
      color: '#999',
    },
    payButton: {
      width: '100%',
      padding: 8,
      borderRadius: 8,
      alignItems: 'center',
      height: 35,
      justifyContent: 'center',
    },
    payButtonText: {
      color: '#fff',
      fontWeight: '600',
      fontSize: 14,
      lineHeight: 14,
    },
    buttonContainer: {
      flexDirection: 'column',
      gap: 10,
      marginTop: 8,
    },
    venmoWrapper: {
      width: '100%',
      marginTop: 0,
    },
  });

export default BillCard;