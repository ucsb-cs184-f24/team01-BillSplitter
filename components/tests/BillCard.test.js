import React from 'react';
import renderer from 'react-test-renderer';
import BillCard from '../BillCard';

describe('<BillCard />', () => {
  const mockItem = {
    id: '1',
    billTitle: 'Lunch',
    billDescription: 'Lunch with friends',
    amount: 25.0,
    category: 'food',
    isCreator: false,
    status: 'pending',
    creatorName: 'John Doe',
    createdAt: new Date(),
  };

  it('renders correctly', () => {
    const tree = renderer.create(
      <BillCard
        item={mockItem}
        currentUser="testUser"
        onPress={() => {}}
        onPaymentPress={() => {}}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('contains all expected children', () => {
    const tree = renderer.create(
      <BillCard
        item={mockItem}
        currentUser="testUser"
        onPress={() => {}}
        onPaymentPress={() => {}}
      />
    ).toJSON();
    expect(tree.children).toBeDefined(); // Ensure it has children
    expect(tree.children.length).toBeGreaterThan(0); // Adjust this as necessary
  });
});
