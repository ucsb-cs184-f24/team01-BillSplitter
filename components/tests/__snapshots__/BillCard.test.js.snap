// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BillCard /> renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    style={
      [
        {
          "borderRadius": 12,
          "borderWidth": 1,
          "flexDirection": "row",
          "marginBottom": 15,
          "padding": 15,
        },
        {
          "backgroundColor": "#F9731615",
        },
        {
          "borderColor": "#F97316",
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "borderRadius": 20,
            "height": 40,
            "justifyContent": "center",
            "marginRight": 12,
            "width": 40,
          },
          {
            "backgroundColor": "#F9731630",
          },
        ]
      }
    >
      <View
        color="#F97316"
        name="coffee"
        size={24}
      />
    </View>
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 8,
          }
        }
      >
        <Text
          style={
            {
              "flex": 1,
              "fontSize": 18,
              "fontWeight": "bold",
              "marginRight": 8,
            }
          }
        >
          Lunch
        </Text>
        <Text
          style={
            [
              {
                "fontSize": 18,
                "fontWeight": "bold",
              },
              {
                "color": "#FF3B30",
              },
            ]
          }
        >
          $
          25.00
        </Text>
      </View>
      <Text
        numberOfLines={2}
        style={
          {
            "color": "#666",
            "fontSize": 14,
            "marginBottom": 8,
          }
        }
      >
        Lunch with friends
      </Text>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 4,
          }
        }
      >
        <Text
          style={
            {
              "color": "#666",
              "fontSize": 14,
            }
          }
        >
          You owe
           
          John Doe
        </Text>
        <Text
          style={
            [
              {
                "fontWeight": "600",
              },
              {
                "color": "#FF3B30",
              },
            ]
          }
        >
          You owe
        </Text>
      </View>
      <Text
        style={
          {
            "color": "#999",
            "fontSize": 12,
          }
        }
      >
        11/20/2024
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#F97316",
            "borderRadius": 8,
            "marginTop": 10,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#fff",
              "fontWeight": "600",
            }
          }
        >
          Pay Now
        </Text>
      </View>
    </View>
  </View>
</View>
`;
